### PR TITLE
Add timeout and retry config options for SNMP plugin (#1472)

### PR DESCRIPTION
**Pull request by cvicente**
Test details: https://gist.github.com/cvicente/b309f9130a50381e11e5140cd97b1dd8

### CODE REVIEW

File: src/snmp.c [hunk 1]
```diff
@@ -449,6 +451,48 @@ static int csnmp_config_add_host_version(host_definition_t *hd,
   return 0;
 } /* int csnmp_config_add_host_address */
 
+static int csnmp_config_add_host_timeout(host_definition_t *hd,
```
Comment (octo) [inline src/snmp.c]: Please use `cf_util_get_int()` from `"configfile.h"` instead:

```c
if (strcasecmp("Timeout", option->key) == 0)
  status = cf_util_get_int(option, &hd->timeout);
```

Or, better yet, change `host_definition_t.timeout` to type `cdtime_t` and use `cf_util_get_cdtime()`.

File: src/snmp.c [hunk 2]
```diff
@@ -449,6 +451,48 @@ static int csnmp_config_add_host_version(host_definition_t *hd,
...
+    return -1;
+  }
+
+  /* net-snmp library timeout is in microseconds */
+  hd->timeout = timeout * 1000000;
+
+  return 0;
+} /* int csnmp_config_add_host_timeout */
+
+static int csnmp_config_add_host_retries(host_definition_t *hd,
```
Comment (octo) [inline src/snmp.c]: Dito:

```c
if (strcasecmp("Retries", option->key) == 0)
  status = cf_util_get_int(option, &hd->retries);
```

File: src/snmp.c [hunk 3]
```diff
@@ -607,6 +655,10 @@ static int csnmp_config_add_host(oconfig_item_t *ci) {
       status = cf_util_get_string(option, &hd->community);
     else if (strcasecmp("Version", option->key) == 0)
       status = csnmp_config_add_host_version(hd, option);
+    else if (strcasecmp("Timeout", option->key) == 0)
```
Comment (octo) [inline src/snmp.c]: These options need to be added to our documentation before this change can be merged. You need to add it to `src/collectd.conf.pod` and `src/collectd.conf.in`.

**Comment by octo** [comment]
P.S.: Saw your test results too late, but I don't quite understand what that shows. You have 5 retries (6 tries total?), each with a timeout of 5s, so the read callback returning after 30s shows that the timeout works?

**Comment by cvicente** [comment]
@octo that's correct. 5 seconds for the initial query, plus 5 x 5 seconds for each retry, total 30 seconds. And yes, we have a custom-built deb package running in production, and we no longer see timeouts in the logs, nor gaps in our graphs.

**Comment by rpv-tomsk** [comment]
Hi, @cvicente!
Thank you for not stopping working on this!

Confirm, altering of `sess.retries` and `sess.timeout` works for UDP.
It looks like `net-snmp` has 5 retries (5+1 total attempts) with 1s timeout each by default.

But, unfortunately, that does not work for TCP connect timeout. 
(I don't know if anybody uses that at practice, but Collectd allows to configure that).

```
[2017-10-18 02:55:27] plugin_read_thread: Handling `snmp-ttt'.
[2017-10-18 02:55:27] Created new plugin context.
[2017-10-18 02:55:27] Initialization complete, entering read-loop.
[2017-10-18 02:57:34] snmp plugin: host ttt: snmp_sess_open failed: Unknown host (tcp:192.168.0.100) (Connection timed out)
[2017-10-18 02:57:34] read-function of plugin `snmp-ttt' failed. Will suspend it for 20.000 seconds.
[2017-10-18 02:57:34] plugin_read_thread: read-function of the `snmp-ttt' plugin took 127.224 seconds, which is above its read interval (20.000 seconds). You might want to adjust the `Interval' or `ReadThreads' settings.
[2017-10-18 02:57:34] plugin_read_thread: read-function of the `snmp-ttt' plugin took 127.224054 seconds.
[2017-10-18 02:57:34] plugin_read_thread: Effective interval of the `snmp-ttt' plugin is 20.000 seconds.
[2017-10-18 02:57:34] plugin_read_thread: Next read of the `snmp-ttt' plugin at 1508270254.073.
[2017-10-18 02:57:34] plugin_read_thread: Handling `snmp-ttt'.
^C[2017-10-18 02:58:49] Exiting normally.
[2017-10-18 02:58:49] collectd: Stopping 5 read threads.
[2017-10-18 02:58:49] plugin: stop_read_threads: Signalling `read_cond'
[2017-10-18 02:59:41] snmp plugin: host ttt: snmp_sess_open failed: Unknown host (tcp:192.168.0.100) (Connection timed out)
[2017-10-18 02:59:41] read-function of plugin `snmp-ttt' failed. Will suspend it for 40.000 seconds.
[2017-10-18 02:59:41] plugin_read_thread: read-function of the `snmp-ttt' plugin took 127.232 seconds, which is above its read interval (40.000 seconds). You might want to adjust the `Interval' or `ReadThreads' settings.
[2017-10-18 02:59:41] plugin_read_thread: read-function of the `snmp-ttt' plugin took 127.231994 seconds.
[2017-10-18 02:59:41] plugin_read_thread: Effective interval of the `snmp-ttt' plugin is 40.000 seconds.
[2017-10-18 02:59:41] plugin_read_thread: Next read of the `snmp-ttt' plugin at 1508270381.305.
[2017-10-18 02:59:41] snmp plugin: Destroying host definition for host `ttt'.
[2017-10-18 02:59:41] collectd: Stopping 5 write threads.
[2017-10-18 02:59:41] plugin: stop_write_threads: Signalling `write_cond'
[2017-10-18 02:59:41] snmp plugin: Destroying all data definitions.
```

Of course, that should not stop us from adding this PR into Collectd.

File: src/snmp.c [hunk 4]
```diff
@@ -449,6 +451,48 @@ static int csnmp_config_add_host_version(host_definition_t *hd,
...
+  }
+
+  timeout = (int)ci->values[0].value.number;
+  if (timeout < 0) {
+    WARNING("snmp plugin: `Timeout' must not be negative");
+    return -1;
+  }
+
+  /* net-snmp library timeout is in microseconds */
+  hd->timeout = timeout * 1000000;
```
Comment (octo) [inline src/snmp.c]: Actually, with this: definitely change `hd->timeout` to be a `cdtime_t` and use `cf_util_get_cdtime()` to parse the config. You can then use `CDTIME_T_TO_US(hd->timeout)` to convert it to the µs expected by net-snmp.

File: src/snmp.c [hunk 5]
```diff
@@ -597,6 +599,10 @@ static int csnmp_config_add_host(oconfig_item_t *ci) {
   hd->sess_handle = NULL;
   hd->interval = 0;
 
+  /* These mean that we have not set a timeout or retry value */
+  hd->timeout = 0xFFFFFFFFFFFFFFFF;
```
Comment (octo) [inline src/snmp.c]: The "invalid" value for `cdtime_t` is zero (`0`), please use that instead.

Comment (cvicente) [reply src/snmp.c]: @octo the issue is that the user may set Timeout to 0 in the config (net-snmp default timeout is 1 sec), and I'm trying to decide if I need to set sess->timeout or not. Suggestions?

Comment (octo) [reply src/snmp.c]:
```c
if (hd->timeout != 0) {
  sedd->timeout = CDTIME_T_TO_US(hd->timeout);
}
```

A timeout of zero doesn't make any sense, in that case just leave `sess` unchanged and go with Net-SNMP's default.

Comment (cvicente) [reply src/snmp.c]: Sounds good. Thanks.

**Pull request merged**